src/common/CArray.hpp

```diff
@@ -156,6 +158,86 @@ struct CArray
     return m_data[ CArrayHelper::linearIndexHelper< DIMS... >::eval( indices ... ) ];
   }
 
+  /**
+   * @struct SubArrayHelper
+   * @brief This struct is used to help with the creation of subarrays.
+   */
+  struct SubArrayHelper
+  {
+    /**
+     * This alias is used to create a const subarray type.
+     */
+    template< int ... SUB_DIMS >
+    using const_type = CArray< T const, T const * const, SUB_DIMS... >;
+
+    /**
+     * This alias is used to create a subarray type.
+     */
+    template< int ... SUB_DIMS >
+    using type = CArray< T, T * const, SUB_DIMS... >;
+  };
+
+  /**
+   * @brief operator[] to access the data in the array or to slice a multidimensional array.
+   * @param index The index to access.
+   * @return A reference to the data at the specified index or the subarray.
+   */
+  SHIVA_CONSTEXPR_HOSTDEVICE_FORCEINLINE
```

Comment:
instead of implementing mutable and immutable accessors with basically the same body, could we implement the mutable one, then in the immutable accessor just call the mutable with a type promotion to const, or am i crazy?

this works fine too, just less code to maintain the other way.

Reply:
done